### PR TITLE
fix(curriculum): add test to exercise tracker

### LINKED ISSUE #44806: Exercise Tracker Challenge - Code with incorrect functionality passes tests
## Affected page

https://www.freecodecamp.org/learn/back-end-development-and-apis/back-end-development-and-apis-projects/exercise-tracker

### The backend route that passes FCC tests despite not working correctly (see comments)
```js
app.get("/api/users/:_id/logs", (req, res) => {
  // get user id from params and check that it won't break the DB query
  const { _id } = req.params;
  if (_id.length !== 24) {
    return res.json({ error: "User ID needs to be 24 hex characters" });
  }

  // find the user
  getUserByIdAnd(_id, (userObject) => {
    if (userObject === null) res.json({ error: "User not found" });
    else {
      const limit = req.query.limit ? req.query.limit : 0;

      // /!\ NOTE `limit` is being applied here BEFORE `from` and `to`
      let promise = ExerciseActivity.find({ user_id: _id }).limit(limit).exec();
      assert.ok(promise instanceof Promise);
      promise.then((exerciseObjects) => {
        // /!\ NOTE `limit` has already been applied at this point, so only
        // the truncated array of exercises will be filtered by `from` and `to`
        if (req.query.from) {
          const from = new Date(req.query.from);
          exerciseObjects = exerciseObjects.filter(
            (e) => new Date(e.date).getTime() >= from.getTime()
          );
        }
        if (req.query.to) {
          const to = new Date(req.query.to);
          exerciseObjects = exerciseObjects.filter(
            (e) => new Date(e.date).getTime() <= to.getTime()
          );
        }
        exerciseObjects = exerciseObjects.map((e) => ({
          ...e,
          date: new Date(e.date).toDateString(),
        }));

        res.json({
          _id: userObject._id,
          username: userObject.username,
          count: exerciseObjects.length,
          log: exerciseObjects,
        });
      });
    }
  });
});
```

#### All exercises for this user
`GET /api/users/61e5689b039759356d4f6a51/logs`
```json
{
  "_id": "61e5689b039759356d4f6a51",
  "username": "cehrlich",
  "count": 6,
  "log": [
    {
      "description": "Jogging",
      "duration": 30,
      "date": "Wed Jan 12 2022"
    },
    {
      "description": "Weights",
      "duration": 60,
      "date": "Mon Jan 11 2021"
    },
    {
      "description": "Climbing",
      "duration": 25,
      "date": "Tue Jan 11 2022"
    },
    {
      "description": "Tennis",
      "duration": 45,
      "date": "Wed Dec 15 2021"
    },
    {
      "description": "Basketball",
      "duration": 25,
      "date": "Sat Feb 12 2022"
    },
    {
      "description": "Swimming",
      "duration": 15,
      "date": "Thu Jan 13 2022"
    }
  ]
}
```

#### Exercises for this user from January 2022 - Note there are three of them
`GET /api/users/61e5689b039759356d4f6a51/logs?from=2022-01-01&to=2022-01-31`
```json
{
  "_id": "61e5689b039759356d4f6a51",
  "username": "cehrlich",
  "count": 3,
  "log": [
    {
      "description": "Jogging",
      "duration": 30,
      "date": "Wed Jan 12 2022"
    },
    {
      "description": "Climbing",
      "duration": 25,
      "date": "Tue Jan 11 2022"
    },
    {
      "description": "Swimming",
      "duration": 15,
      "date": "Thu Jan 13 2022"
    }
  ]
}
```

#### Limiting the results to 2 only returns 1 - This is because the query was limited to the first two objects ("Jogging" and "Weights") _before_ the `from` and `to` filters are applied
`GET /api/users/61e5689b039759356d4f6a51/logs?from=2022-01-01&to=2022-01-31&limit=2`
```json
{
  "_id": "61e5689b039759356d4f6a51",
  "username": "cehrlich",
  "count": 1,
  "log": [
    {
      "description": "Jogging",
      "duration": 30,
      "date": "Wed Jan 12 2022"
    }
  ]
}
```

## Expected behavior

The Challenge requirements state:
> You can add from, to and limit parameters to a GET /api/users/:_id/logs request to retrieve part of the log of any user. from and to are dates in yyyy-mm-dd format. limit is an integer of how many logs to send back.

The above code / functionality should not pass the test.

## System

(this is an issue with the tests so it will be the same for any client device, but I'm using:)
- Device: MacBook Air (Late 2020)
- OS: MacOS 12.1
- Browser: Chrome
- Version: 97

## Additional context

If you believe the tests should catch this, I would be happy to rewrite the relevant test and create a Pull Request for it.

### PR DESCRIPTION
**Pull request by c-ehrlich**
Checklist:

- [x] I have read [freeCodeCamp's contribution guidelines](https://contribute.freecodecamp.org).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

Closes #44806

* Add a test to prevent incorrectly functioning code from passing the Exercise Tracker project requirements - see Issue#44806 for more details.
* Date of addExerciseTwoRes was changed to allow a one day buffer on either side of the new request (the existing tests have this also)
* I tried to choose a descriptive name for the new res - let me know if something shorter is preferred.

### CODE REVIEW

File: curriculum/challenges/english/05-back-end-development-and-apis/back-end-development-and-apis-projects/exercise-tracker.md [hunk 1]
```diff
@@ -591,6 +591,16 @@ async (getUserInput) => {
       } else {
         throw new Error(`${limitRes.status} ${limitRes.statusText}`);
       }
+      const filterDateBeforeLimitRes = await fetch(
+        url + `/api/users/${_id}/logs/from=1990-01-02&to=1990-01-04&limit=1`
```
Comment (ShaunSHamilton) [inline curriculum/challenges/english/05-back-end-development-and-apis/back-end-development-and-apis-projects/exercise-tracker.md:595]:
```suggestion
        url + `/api/users/${_id}/logs?from=1990-01-02&to=1990-01-04&limit=1`
```

**Review by ShaunSHamilton — CHANGES_REQUESTED** [review]
Thank you, for taking the time to contribute.

Just a small fix

**Review by ShaunSHamilton — APPROVED** [review]
Thanks for this.

This LGTM 👍

**Review by moT01 — APPROVED** [review]
LGTM @c-ehrlich 🎉 Thanks for fixing this and congratulations on your first PR for the freeCodeCamp repo 🎉 We welcome future contributions 👍

**Pull request merged**